plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs

```diff
@@ -93,51 +93,69 @@ becomes
 
-strictNonFunctionBinding :: Binding tyname name uni fun a -> Bool
-strictNonFunctionBinding = \case
-    TermBind _ Strict (VarDecl _ _ ty) _ | not $ isFunctionType ty -> True
-    _                                                              -> False
-
-nonStrictify :: Binding tyname name uni fun a -> Binding tyname name uni fun a
-nonStrictify = \case
+nonStrictifyB :: Binding tyname name uni fun a -> Binding tyname name uni fun a
+nonStrictifyB = \case
     TermBind x _ d rhs -> TermBind x NonStrict d rhs
     b                  -> b
 
-mkStrictifyingBinding
-    :: (ToBuiltinMeaning uni fun, PLC.HasUnique name PLC.TermUnique)
-    => BuiltinsInfo uni fun
-    -> VarsInfo tyname name
-    -> Binding tyname name uni fun a
-    -> Maybe (Binding tyname name uni fun a)
-mkStrictifyingBinding binfo vinfo = \case
-    -- Only need to strictify if the previous binding was not definitely pure
-    -- Also, we're reusing the same variable here, see Note [Thunking recursions]
-    TermBind x _ d rhs | not (isPure binfo vinfo rhs) -> Just $ TermBind x Strict d (mkVar x d)
-    _                                                 -> Nothing
+-- Out of a binding `(vardecl x = rhs)`, make a "strictifier" binding: `(strict vardecl x = x)`
+mkStrictifierB :: Binding tyname name uni fun a -> Binding tyname name uni fun a
+mkStrictifierB = \case
+    TermBind x _ d _ -> TermBind x Strict d (mkVar x d)
+    b                -> b
 
 thunkRecursionsStep
-    :: (ToBuiltinMeaning uni fun, PLC.HasUnique name PLC.TermUnique)
+    :: forall tyname name uni fun a
+    . (ToBuiltinMeaning uni fun, PLC.HasUnique name PLC.TermUnique)
     => BuiltinsInfo uni fun
     -> VarsInfo tyname name
     -> Term tyname name uni fun a
     -> Term tyname name uni fun a
-thunkRecursionsStep binfo vinfo (Let a Rec bs t) =
+thunkRecursionsStep binfo vinfo = \case
+ -- only apply the transformation if there is at least 1 problematic binding in a letrec group
+  Let a Rec bs t | any isProblematic bs ->
     -- See Note [Thunking recursions]
-    let (toThunk, noThunk) = partition strictNonFunctionBinding bs
-        newBindings = fmap nonStrictify toThunk ++ noThunk
-        strictifiers = mapMaybe (mkStrictifyingBinding binfo vinfo) toThunk
-    in mkLet a Rec newBindings $ mkLet a NonRec strictifiers t
-thunkRecursionsStep _ _ t = t
+    let (toNonStrictify, rest) = NE.partition needsNonStrictify bs
+        -- MAYBE: use some prism/traversal to keep the original arrangement of the let group
+        -- this is not a semantic problem, but just a stylistic/debugging issue where the original
+        -- let-group will have reordered the (lazified) bindings
+        editedLet = mkLet a Rec $ fmap nonStrictifyB toNonStrictify ++ rest
+        -- we now have to consider for which of those non-strictified we are going to introduce strictifiers
+        -- to preserve the original strictness annotations, we have to introduce strictifiers *for all* those we previously non-strictified
```

Comment:
The real reason is to preserve any potential effects, right?

Reply:
Indeed, this comment came from somewhere else. The comment made sense previously but not now. Fixing